File: Cpp/Binary Search.cpp
class Solution {
public:
    /**
     * @param nums: The integer array.
     * @param target: Target number to find.
     * @return: The first position of target. Position starts from 0.
     */
    int binarySearch(vector<int> &array, int target) {
        // write your code here
        int left = 0, right = array.size() - 1, mid;
        while (left <= right) {
          mid = (left + right) / 2;
          if (array[mid] == target) {
            if(mid > 0 && array[mid - 1] == target) right = mid - 1;
            else return mid;
          }
          else if (array[mid] > target) {
            right = mid - 1;
          }
          else left = mid + 1;
        }
        return -1;
    }
};

File: Java/Binary-Search.java
class Solution {
    /**
     * @param nums: The integer array.
     * @param target: Target to find.
     * @return: The first position of target. Position starts from 0.
     */
    public int binarySearch(int[] nums, int target) {
        //write your code here
        int left = 0, right = nums.length, mid;
        while (left + 1 < right) {
        	mid = left + (right - left) / 2;
        	if (nums[mid] == target) {
        		right = mid;
        	} else if (nums[mid] > target) {
        		right = mid;
        	} else {
        		left = mid;
        	}
        }
        if (nums[left] == target) return left;
        else if(nums[right] == target) return right;
        return -1;
    }
}
